src/stdlib/Polynetwork.scillib

```diff
@@ -0,0 +1,674 @@
+    let buff = append_uint16 buff m in
+    let buff_sha256 = builtin sha256hash buff in
+    let nextBookKeeper = builtin ripemd160hash buff_sha256 in
+
+    Pair {ByStr20 (List ByStr20)} nextBookKeeper keepers
+
+let lengther_pubkey = @list_length Pubkey
+let lengther_address = @list_length ByStr20
+
+let compute_m : Uint32 -> Uint32 =
+  fun (n : Uint32) =>
+    let one = Uint32 1 in
+    let t1 = builtin sub n one in
+    let three = Uint32 3 in
+    let t2 = builtin div t1 three in
+    let m = builtin sub n t2 in
+    m
+
+(* See EthCrossChainUtils.sol:verifyPubKey *)
+let verifyPubkey : List Pubkey -> Pair ByStr20 (List ByStr20) =
+  fun (pubKeys : List Pubkey) =>
+    let n = lengther_pubkey pubKeys in
+    let m = compute_m n in
+    getBookKeeper m pubKeys
+
+(* Mimics ethereum's recovery of address from message and a signature. *)
+let ecrecover : ByStr -> ByStr64 -> Uint32 -> ByStr20 =
+  fun (msg : ByStr) =>
+  fun (sig : ByStr64) =>
+  fun (recid : Uint32) =>
+    let pk = builtin ecdsa_recover_pk msg sig recid in
+    let pos = Uint32 1 in
+    let len = Uint32 64 in
+    let pk_bs = builtin to_bystr pk in
+    let pk_ = builtin substr pk_bs pos len in
+    let pkHash = builtin keccak256hash pk_ in
+    let pkHash_bs = builtin to_bystr pkHash in
+    let pos = Uint32 12 in
+    let len = Uint32 20 in
+    let addr_bs = builtin substr pkHash_bs pos len in
+    let addr_opt = builtin to_bystr20 addr_bs in
+    let sommer = @some_exn ByStr20 in
+    let exnobj = 0x0000000000000000000000000000000000000000 in
+    sommer addr_opt exnobj
+
+(* See EthCrossChainUtils.sol:verifySig *)
+
+let verifySig : ByStr -> List Signature -> List ByStr20 -> Uint32 -> Bool =
+  fun (rawHeader : ByStr) =>
+  fun (siglist : List Signature) =>
+  fun (keepers : List ByStr20) =>
+  fun ( m : Uint32) =>
+    let header_hash =
+      fun (header : ByStr) =>
+        let h1 = builtin sha256hash header in
+        let h2 = builtin sha256hash h1 in
+        builtin to_bystr h2
+    in
+    let hashedHeader = header_hash rawHeader in
+    let mapf =
+      fun (s : Signature) =>
+        match s with
+        | Signature rs_v =>
+          let rs_v = builtin to_bystr rs_v in
+          let pos = Uint32 64 in
+          let len = Uint32 1 in
+          let v_bs = builtin substr rs_v pos len in
+          let v_bsx_opt = builtin to_bystr1 v_bs in
+          let exnobj_bystr1 = 0x00 in
+          let sommer_bystr1 = @some_exn ByStr1 in
+          let v_bsx = sommer_bystr1 v_bsx_opt exnobj_bystr1 in
+          let v = builtin to_uint32 v_bsx in
+          let pos = Uint32 0 in
+          let len = Uint32 64 in
+          let rs_bs = builtin substr rs_v pos len in
+          let rs_bs64_opt = builtin to_bystr64 rs_bs in
+          let exnobj_bystr64 = 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 in
+          let sommer_bystr64 = @some_exn ByStr64 in
```

Comment:
`slice` is just substring computation, yes.

Reply:
So this computation here is separating the signature into two `(r,s)` components, and then the last byte is the `v` component. We don't do the `+ 27` though because that is an EVM specific thing. Our `ecrecover` doesn't need it.

Reply:
I had to do a lot of trial-and-error to get this right. Finally this helped: https://ethereum.stackexchange.com/questions/13778/get-public-key-of-any-ethereum-account

Reply:
I added this reference into the code now.